src/GenLevelStudiesHists.cc

```diff
@@ -20,47 +21,103 @@ using namespace std;
-  int id2 = event.geninfo->id1();
-  hist<TH1F>("sumweights")->Fill(1, weight);
+  // int id1 = event.geninfo->id1();
+  // int id2 = event.geninfo->id1();
+  // hist<TH1F>("sumweights")->Fill(1, weight);
+  //
+  // hist<TH1F>("geninfoscalepdf")->Fill(event.geninfo->scale_pdf(), weight);
+  // hist<TH1F>("geninfoid1")->Fill(id1, weight);
+  // hist<TH1F>("geninfoid2")->Fill(id2, weight);
+  // hist<TH1F>("geninfox1")->Fill(event.geninfo->x1(), weight);
+  // hist<TH1F>("geninfox2")->Fill(event.geninfo->x2(), weight);
+  //
+  //
+  // if(abs(id1) == 21) {
+  //   hist<TH1F>("geninfoxg")->Fill(event.geninfo->x1(), weight);
+  //   hist<TH1F>("geninfoxg")->Fill(event.geninfo->x2(), weight);
+  // }
+  // if(abs(id1) >= 1 && abs(id1) <= 3) {
+  //   hist<TH1F>("geninfoxlight")->Fill(event.geninfo->x1(), weight);
+  //   hist<TH1F>("geninfoxlight")->Fill(event.geninfo->x2(), weight);
+  // }
+  //
+  // hist<TH1F>("genmet")->Fill(event.genmet->pt(), weight);
+  // hist<TH1F>("genmet_phi")->Fill(event.genmet->phi(), weight);
+  // hist<TH2D>("genmet_phi_pt")->Fill(event.genmet->phi(), event.genmet->pt(), weight);
 
-  hist<TH1F>("geninfoscalepdf")->Fill(event.geninfo->scale_pdf(), weight);
-  hist<TH1F>("geninfoid1")->Fill(id1, weight);
-  hist<TH1F>("geninfoid2")->Fill(id2, weight);
-  hist<TH1F>("geninfox1")->Fill(event.geninfo->x1(), weight);
-  hist<TH1F>("geninfox2")->Fill(event.geninfo->x2(), weight);
+  hist<TH1F>("NumberOfParticles")->Fill(event.genparticles_all->size(),weight);
 
+  int NChargedParticles = 0;
+  TLorentzVector FourLeptons;                                   // Initialise counter
+  for(size_t i=0; i<event.genparticles_all->size(); i++){
+    GenParticle m = event.genparticles_all->at(i);
+    // hist<TH1F>("PdgId")->Fill(m.pdgid(), weight);
+    hist<TH1F>("PdgId2")->Fill(m.pdgid(), weight);
 
-  if(abs(id1) == 21) {
-    hist<TH1F>("geninfoxg")->Fill(event.geninfo->x1(), weight);
-    hist<TH1F>("geninfoxg")->Fill(event.geninfo->x2(), weight);
-  }
-  if(abs(id1) >= 1 && abs(id1) <= 3) {
-    hist<TH1F>("geninfoxlight")->Fill(event.geninfo->x1(), weight);
-    hist<TH1F>("geninfoxlight")->Fill(event.geninfo->x2(), weight);
-  }
+    if (m.pdgid() == 25 && m.get_statusflag(GenParticle::isLastCopy))  {
+      if (!m.get_statusflag(GenParticle::isHardProcess) && !m.get_statusflag(GenParticle::fromHardProcessBeforeFSR)) continue;
+      hist<TH1F>("verifH")->Fill(m.m(), weight);
+      for(size_t j=0; j<event.genparticles_all->size(); j++) {
+        GenParticle z = event.genparticles_all->at(j);
+        if (z.mother_identifier() == m.identifier()) {
+          hist<TH1F>("verifZ")->Fill(z.m(), weight);
+          for(size_t k=0; k<event.genparticles_all->size(); k++) {
+            GenParticle l = event.genparticles_all->at(k);
+            if (l.mother_identifier() == z.identifier()) {
+              FourLeptons += l.p4();
+            }
+          }
+        }
+      }
+    }
+    TLorentzVector zero;
```

Comment:
This is ok. Another way is to fill a vector with the identifier or index of the 4L and then create the TLorentzVector and fill the hists if there are exactly 4 indices.